halo/attest/voter/voter.go

```diff
@@ -63,38 +78,121 @@ func LoadVoter(ctx context.Context, privKey crypto.PrivKey, path string, provide
 		return nil, err
 	}
 
-	a := Voter{
-		privKey: privKey,
-		address: addr,
-		path:    path,
-		chains:  chains,
+	return &Voter{
+		privKey:  privKey,
+		address:  addr,
+		path:     path,
+		chains:   chains,
+		provider: provider,
 
 		available: s.Available,
 		proposed:  s.Proposed,
 		committed: s.Committed,
+	}, nil
+}
+
+// Start starts runners that attest to each source chain. It does not block, it returns immediately.
+func (a *Voter) Start(ctx context.Context, app App) {
+	for chainID := range a.chains {
+		go a.runForever(ctx, app, chainID)
 	}
+}
 
-	// This shouldn't be required, but -race otherwise complains...
-	a.mu.Lock()
-	defer a.mu.Unlock()
+// runForever blocks, repeatedly calling runOnce (with backoff) for the provided chain until the context is canceled.
+func (a *Voter) runForever(ctx context.Context, app App, chainID uint64) {
+	ctx = log.WithCtx(ctx, "chain", a.chains[chainID])
 
-	// Subscribe to latest block provider
-	for chainID := range chains {
-		var fromHeight uint64
-		if latest, ok := a.latestByChainUnsafe(chainID); ok {
-			fromHeight = latest.BlockHeader.Height + 1
+	backoff := expbackoff.New(ctx, expbackoff.WithPeriodicConfig(time.Second))
+	for ctx.Err() == nil {
+		if !app.IsValidator(ctx, a.address) {
+			backoff()
+			continue
 		}
-		err := provider.Subscribe(ctx, chainID, fromHeight, a.Attest)
-		if err != nil {
-			return nil, err
+
+		err := a.runOnce(ctx, app, chainID)
+		if ctx.Err() != nil {
+			return // Don't log or sleep on context cancel.
 		}
+
+		log.Warn(ctx, "Vote runner failed (will retry)", err)
+		backoff()
+	}
+}
+
+// runOnce blocks, streaming xblocks from the provided chain until an error is encountered.
+// It always returns a non-nil error.
+func (a *Voter) runOnce(ctx context.Context, app App, chainID uint64) error {
```

Comment:
Why always return a non-nil error?

Reply:
because it returns only when:
 - a) it encounters an error (it returns that error)
 - b) the context is cancelled (it return ErrContextCancelled)